### ISSUE TITLE
Question about GRPO and sample indexing

### ISSUE THREAD

**Issue by scris** [PeterGriffinJin/Search-R1#88]
I see fix for [GRPO sample indexing](https://github.com/PeterGriffinJin/Search-R1/commit/9ec2fa9892fbf0315d0c67b4dc08ae8f6cf5f378) is included in v0.2. 

And for v0.1, GRPO can work well as recorded in [Search-r1-grpo-qwen2.5-3b-it-em](https://wandb.ai/peterjin/Search-R1-nq_hotpotqa_train/runs/wg54cbbs).

However, in v0.2, [Search-r1-grpo-qwen2.5-7b-it-em](https://wandb.ai/peterjin/Search-R1-v0.2/runs/klposne8) seems to be somehow wrong as the experiment exploded up in around step 160.

Does that mean we need to conduct other changes to fix GRPO? I'm asking as I also find this as a bug and fixed it, and my experiment then also will explode somewhere. Thanks.

By the way, do you think this can be fixed by solution in https://github.com/0russwest0/Agent-R1/issues/30? Thanks.

**Comment by PeterGriffinJin** [comment]
Hi there,

Thank you for your interest in our work. We have a similar observation on GRPO collapse and would like to share some of our findings and thoughts

- And for v0.1, GRPO can work well as recorded in [Search-r1-grpo-qwen2.5-3b-it-em](https://wandb.ai/peterjin/Search-R1-nq_hotpotqa_train/runs/wg54cbbs).

For v0.1 implementation, there is a small error on the group indexing (https://github.com/PeterGriffinJin/Search-R1/commit/9ec2fa9892fbf0315d0c67b4dc08ae8f6cf5f378). The same prompts are not given the same ID so GRPO just degrades to REINFORCE (group size = 1). In this case, we find the training is more stable than GRPO.

- However, in v0.2, [Search-r1-grpo-qwen2.5-7b-it-em](https://wandb.ai/peterjin/Search-R1-v0.2/runs/klposne8) seems to be completely wrong.

Yes, we also have a similar observation that GRPO with a larger group size would converge faster and collapse earlier. Finally, the LLM would just output '!!!!!' as responses. We have tried removing the KL loss and entropy loss during GRPO but find it still collapses.

- Does that mean we need to conduct other changes to fix GRPO? I'm asking as I also find this as a bug and fixed it, and my experiment then also will explode somewhere. Thanks.

I am not sure if this is a bug or the inherent unstable nature of the RL training. In contrast, we do find PPO/REINFORCE to be more stable than GRPO. Please feel free to let us know if you identify any potential bug which may lead to this collapse for GRPO.

Hope that helps.

**Comment by scris** [comment]
Thanks for your kind reply :)

**Comment by Yux1angJi** [comment]
Hi, I also meet model collapse during GRPO training.
The entropy loss and kl loss come to nan and the model just output !!!! after collapse.
Seems that this is caused by big gap between log_prob and old_log_prob, and lead to nan value finally.
```
negative_approx_kl = log_prob - old_log_prob
ratio = torch.exp(negative_approx_kl)
```

I see that the recent verl version will add the clamp operation here. I would like to ask if such instability will still occur in the newest integrated Search-R1 verl version. Because I did not find collapse at released [log](https://wandb.ai/lingchang-ustc/search_async_rl/workspace?nw=nwuserlingchang).

**Comment by xjtupy** [comment]
> Hi there, I've tried to add the clamp before ratio calculation. However, it seems that it still collapse TAT

I also have this problem, has it been solved?

**Comment by xjtupy** [comment]
> Hi there,
> 
> Thank you for pointing this out! I guess the integrated Search-R1 verl version could be more stable. Feel free to try it out and let use know.
> 
> It seems that they add this KL clamp before doing the ratio calculation. We will also explore this in our codebase these days and share the results then.
> 
> ```
> negative_approx_kl = torch.clamp(negative_approx_kl, min=-20.0, max=20.0)
> ```

Hi, has this problem occurred in your recent experiments?

**Comment by giantzhong** [comment]
After I addedAfter attempting to add the KL clamp operation, the training did not collapse, and the situation where Entropy_loss, pg_loss, and kl_loss became NAN did not occur as before.
